onnxruntime/core/optimizer/constant_sharing.cc

```diff
@@ -0,0 +1,233 @@
+      if (found_subgraph_usage) {
+        break;
+      }
+
+      // Iterate all input defs to replace those that are equal to origin_initializer_node_arg,
+      // Then it would be safe to remove the consumer node aferwards.
+      for (int i = 0; i < static_cast<int>(const_node->InputDefs().size()); ++i) {
+        if (const_node->InputDefs()[i] == origin_initializer_node_arg) {
+          consumer_node_to_input_index_map[const_node].push_back(i);
+        }
+      }
+    }
+
+    if (found_subgraph_usage || consumer_node_to_input_index_map.size() == 0) {
+      continue;
+    }
+
+    onnxruntime::Initializer initializer{*tensor_proto, graph.ModelPath()};
+    std::int32_t value_id;
+    switch (data_type) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
+        float flt_val = *initializer.data<float>();
+        if (float_value_to_value_id_map.find(flt_val) == float_value_to_value_id_map.end()) {
+          float_value_to_value_id_map[flt_val] = GenerateUniqueValueId();
+        }
+        value_id = float_value_to_value_id_map[flt_val];
+      } break;
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16: {
+        float flt16_val = math::halfToFloat(initializer.data<MLFloat16>()->val);
+        if (half_value_to_value_id_map.find(flt16_val) == half_value_to_value_id_map.end()) {
+          half_value_to_value_id_map[flt16_val] = GenerateUniqueValueId();
+        }
+        value_id = half_value_to_value_id_map[flt16_val];
+      } break;
+      case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE: {
+        double dbl_val = *initializer.data<double>();
+        if (double_value_to_value_id_map.find(dbl_val) == double_value_to_value_id_map.end()) {
+          double_value_to_value_id_map[dbl_val] = GenerateUniqueValueId();
+        }
+        value_id = double_value_to_value_id_map[dbl_val];
+      } break;
+      case ONNX_NAMESPACE::TensorProto_DataType_INT32: {
+        int32_t int32_val = *initializer.data<int32_t>();
+        if (int32_value_to_value_id_map.find(int32_val) == int32_value_to_value_id_map.end()) {
+          int32_value_to_value_id_map[int32_val] = GenerateUniqueValueId();
+        }
+        value_id = int32_value_to_value_id_map[int32_val];
+      } break;
+      case ONNX_NAMESPACE::TensorProto_DataType_INT64: {
+        int64_t int64_val = *initializer.data<int64_t>();
+        if (int64_value_to_value_id_map.find(int64_val) == int64_value_to_value_id_map.end()) {
+          int64_value_to_value_id_map[int64_val] = GenerateUniqueValueId();
+        }
+        value_id = int64_value_to_value_id_map[int64_val];
+      } break;
+      default:
+        ORT_THROW("Unsupported data type [",
+                  std::to_string(data_type) + "] found: for initializer: " + initializer_name);
+    }
+
+    // Construct a string by data type, value, and rank. Used as a key in pattern_key_to_shared_arg_map.
+    const std::string& pattern_key = MakeString(SHARED_INITIALIZER_PREFIX, value_id, "_", data_type, "_",
```

Comment:
just trying to understand how this can be a string ref? Does this function generate a string and return it? Or does create a static string and return a ref?

Reply:
Do you think maybe this was causing the access violation?

Reply:
Shouldn't be a ref.

Reply:
Sorry, it's my mistake. The ref is newly added in recent commit, should not related to the previous access violation.